src/category_theory/bundled_hom.lean

```diff
@@ -0,0 +1,103 @@
+/-
+Copyright (c) 2019 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+-/
+import category_theory.bundled
+import category_theory.types
+
+/-!
+# Category instances for algebraic structures that use bundled homs.
+
+Many algebraic structures in Lean initially used unbundled homs (e.g. a bare function between types, along with
+an a `is_monoid_hom` typeclass), but the general trend is towards using bundled homs.
+
+While the helper functions in `category_theory/concrete_category.lean` are useful for categories
+with unbundled homs, this file provides similar infrastructure for categories with bundled homs.
+-/
+
+universes w v u
+
+namespace category_theory
+
+variables (c : Type u → Type v)
+
+-- TODO the square brackets here are useless, remove them and the resulting @s?
+structure bundled_hom :=
+(hom : Π {α β : Type u} (Iα : c α) (Iβ : c β), Type w)
+(to_fun : Π {α β : Type u} (Iα : c α) (Iβ : c β), hom Iα Iβ → α → β)
+(id : Π {α : Type u} (I : c α), hom I I)
+(comp : Π {α β γ : Type u} (Iα : c α) (Iβ : c β) (Iγ : c γ),
+  hom Iα Iβ → hom Iβ Iγ → hom Iα Iγ)
+(hom_ext : Π {α β : Type u} (Iα : c α) (Iβ : c β) {f g : hom Iα Iβ}
+  (h : to_fun Iα Iβ f = to_fun Iα Iβ g), f = g . obviously)
+(id_to_fun : Π {α : Type u} (I : c α), to_fun I I (id I) = _root_.id . obviously)
+(comp_to_fun : Π {α β γ : Type u} (Iα : c α) (Iβ : c β) (Iγ : c γ)
+  (f : hom Iα Iβ) (g : hom Iβ Iγ),
+  to_fun Iα Iγ (comp Iα Iβ Iγ f g) = (to_fun Iβ Iγ g) ∘ (to_fun Iα Iβ f) . obviously)
+
+attribute [class] bundled_hom
+
+attribute [extensionality] bundled_hom.hom_ext
+attribute [simp] bundled_hom.id_to_fun bundled_hom.comp_to_fun
+
+namespace bundled_hom
+
+section
+variable [S : bundled_hom c]
+include S
+
+instance : category (bundled c) :=
+{ hom := λ X Y, @bundled_hom.hom c S X.α Y.α X.str Y.str,
+  id := λ X, @bundled_hom.id c S X.α X.str,
+  comp := λ X Y Z f g, @bundled_hom.comp c S X.α Y.α Z.α X.str Y.str Z.str f g }
+
+instance {X Y : bundled c} : has_coe_to_fun (X ⟶ Y) :=
```

Comment:
This way we have to coercions from `M →* N` to `M → N`. Can it lead to some troubles like “I can't rewrite using this lemma, because it uses a different `has_coe_to_fun` instance”?

Reply:
Yes, I agree this look like a problem. Does your approach avoid this?

Reply:
One possible approach is to make it a local instance. Then we can use it in the generic code but not outside.

Reply:
Hmm.. it does get used in the file about `cones`. For now I'll make it a local instance, and add it where necessary.

Reply:
Seems okay.